File: frontend/src/store/services/dataDriver.ts
import { api } from './api'

// 数据源相关接口
export interface DataSource {
  id: number
  name: string
  description?: string
  type: string
  project_id?: number
  config?: Record<string, any>
  is_active?: boolean
  created_by?: number
  created_at?: string
  updated_at?: string
}

export interface DataSourceCreate {
  name: string
  description?: string
  type: string
  project_id?: number
  config?: Record<string, any>
  is_active?: boolean
}

export interface DataSourceUpdate {
  name?: string
  description?: string
  type?: string
  project_id?: number
  config?: Record<string, any>
  is_active?: boolean
}

// 数据模板相关接口
export interface DataTemplate {
  id: number
  name: string
  description?: string
  data_source_id: number
  project_id?: number
  mapping?: Record<string, any>
  filters?: Record<string, any>
  loop_strategy?: string
  created_by?: number
  created_at?: string
  updated_at?: string
}

export interface DataTemplateCreate {
  name: string
  description?: string
  data_source_id: number
  project_id?: number
  mapping?: Record<string, any>
  filters?: Record<string, any>
  loop_strategy?: string
}

export interface DataTemplateUpdate {
  name?: string
  description?: string
  data_source_id?: number
  project_id?: number
  mapping?: Record<string, any>
  filters?: Record<string, any>
  loop_strategy?: string
}

// 数据生成器相关接口
export interface DataGenerator {
  id: number
  name: string
  description?: string
  project_id?: number
  type: string
  config?: Record<string, any>
  is_active?: boolean
  created_by?: number
  created_at?: string
  updated_at?: string
}

export interface DataGeneratorCreate {
  name: string
  description?: string
  project_id?: number
  type: string
  config?: Record<string, any>
  is_active?: boolean
}

export interface DataGeneratorUpdate {
  name?: string
  description?: string
  project_id?: number
  type?: string
  config?: Record<string, any>
  is_active?: boolean
}

export interface DataDriverListParams {
  project_id?: number
  type?: string
  is_active?: boolean
  search?: string
  skip?: number
  limit?: number
}

export const dataDriverService = {
  // 数据源相关API
  async getDataSources(params?: DataDriverListParams): Promise<DataSource[]> {
    const response = await api.get<DataSource[]>('/data-drivers/data-sources', { params })
    return response.data
  },

  async getDataSource(id: number): Promise<DataSource> {
    const response = await api.get<DataSource>(`/data-drivers/data-sources/${id}`)
    return response.data
  },

  async createDataSource(data: DataSourceCreate): Promise<DataSource> {
    const response = await api.post<DataSource>('/data-drivers/data-sources', data)
    return response.data
  },

  async updateDataSource(id: number, data: DataSourceUpdate): Promise<DataSource> {
    const response = await api.put<DataSource>(`/data-drivers/data-sources/${id}`, data)
    return response.data
  },

  async deleteDataSource(id: number): Promise<void> {
    await api.delete(`/data-drivers/data-sources/${id}`)
  },

  // 数据模板相关API
  async getDataTemplates(params?: { project_id?: number; data_source_id?: number; search?: string }): Promise<DataTemplate[]> {
    const response = await api.get<DataTemplate[]>('/data-drivers/data-templates', { params })
    return response.data
  },

  async getDataTemplate(id: number): Promise<DataTemplate> {
    const response = await api.get<DataTemplate>(`/data-drivers/data-templates/${id}`)
    return response.data
  },

  async createDataTemplate(data: DataTemplateCreate): Promise<DataTemplate> {
    const response = await api.post<DataTemplate>('/data-drivers/data-templates', data)
    return response.data
  },

  async updateDataTemplate(id: number, data: DataTemplateUpdate): Promise<DataTemplate> {
    const response = await api.put<DataTemplate>(`/data-drivers/data-templates/${id}`, data)
    return response.data
  },

  async deleteDataTemplate(id: number): Promise<void> {
    await api.delete(`/data-drivers/data-templates/${id}`)
  },

  // 数据生成器相关API
  async getDataGenerators(params?: DataDriverListParams): Promise<DataGenerator[]> {
    const response = await api.get<DataGenerator[]>('/data-drivers/data-generators', { params })
    return response.data
  },

  async getDataGenerator(id: number): Promise<DataGenerator> {
    const response = await api.get<DataGenerator>(`/data-drivers/data-generators/${id}`)
    return response.data
  },

  async createDataGenerator(data: DataGeneratorCreate): Promise<DataGenerator> {
    const response = await api.post<DataGenerator>('/data-drivers/data-generators', data)
    return response.data
  },

  async updateDataGenerator(id: number, data: DataGeneratorUpdate): Promise<DataGenerator> {
    const response = await api.put<DataGenerator>(`/data-drivers/data-generators/${id}`, data)
    return response.data
  },

  async deleteDataGenerator(id: number): Promise<void> {
    await api.delete(`/data-drivers/data-generators/${id}`)
  },
}

File: backend/app/schemas/data_driver.py
"""
数据驱动配置相关的 Pydantic 模型
"""
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any
from datetime import datetime


# 数据源相关模型
class DataSourceBase(BaseModel):
    """数据源基础模型"""
    name: str = Field(..., min_length=1, max_length=200, description="数据源名称")
    description: Optional[str] = Field(None, description="描述")
    type: str = Field(..., description="数据源类型")
    project_id: Optional[int] = Field(None, description="项目ID")
    config: Optional[Dict[str, Any]] = Field(None, description="数据源配置")
    is_active: Optional[bool] = Field(default=True, description="是否激活")


class DataSourceCreate(DataSourceBase):
    """创建数据源模型"""
    pass


class DataSourceUpdate(BaseModel):
    """更新数据源模型"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    type: Optional[str] = None
    project_id: Optional[int] = None
    config: Optional[Dict[str, Any]] = None
    is_active: Optional[bool] = None


class DataSourceResponse(DataSourceBase):
    """数据源响应模型"""
    id: int
    created_by: Optional[int] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


# 数据模板相关模型
class DataTemplateBase(BaseModel):
    """数据模板基础模型"""
    name: str = Field(..., min_length=1, max_length=200, description="模板名称")
    description: Optional[str] = Field(None, description="描述")
    data_source_id: int = Field(..., description="关联的数据源ID")
    project_id: Optional[int] = Field(None, description="项目ID")
    mapping: Optional[Dict[str, Any]] = Field(None, description="数据映射配置")
    filters: Optional[Dict[str, Any]] = Field(None, description="数据过滤条件")
    loop_strategy: Optional[str] = Field(default='all', description="循环策略")


class DataTemplateCreate(DataTemplateBase):
    """创建数据模板模型"""
    pass


class DataTemplateUpdate(BaseModel):
    """更新数据模板模型"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    data_source_id: Optional[int] = None
    project_id: Optional[int] = None
    mapping: Optional[Dict[str, Any]] = None
    filters: Optional[Dict[str, Any]] = None
    loop_strategy: Optional[str] = None


class DataTemplateResponse(DataTemplateBase):
    """数据模板响应模型"""
    id: int
    created_by: Optional[int] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


# 数据生成器相关模型
class DataGeneratorBase(BaseModel):
    """数据生成器基础模型"""
    name: str = Field(..., min_length=1, max_length=200, description="生成器名称")
    description: Optional[str] = Field(None, description="描述")
    project_id: Optional[int] = Field(None, description="项目ID")
    type: str = Field(..., description="生成器类型")
    config: Optional[Dict[str, Any]] = Field(None, description="生成器配置")
    is_active: Optional[bool] = Field(default=True, description="是否激活")


class DataGeneratorCreate(DataGeneratorBase):
    """创建数据生成器模型"""
    pass


class DataGeneratorUpdate(BaseModel):
    """更新数据生成器模型"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    project_id: Optional[int] = None
    type: Optional[str] = None
    config: Optional[Dict[str, Any]] = None
    is_active: Optional[bool] = None


class DataGeneratorResponse(DataGeneratorBase):
    """数据生成器响应模型"""
    id: int
    created_by: Optional[int] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True

